plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -424,6 +445,74 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
 	}, nil
 }
 
+// selectNodesForPreemption finds all the nodes with possible victims for
+// preemption in parallel.
+func selectNodesForPreemption(pod *v1.Pod,
+	nodeNameToInfo map[string]*schedulercache.NodeInfo,
+	nodes []*v1.Node,
+	predicates map[string]algorithm.FitPredicate,
+	metadataProducer algorithm.MetadataProducer,
+) map[string][]*v1.Pod {
+
+	nodeNameToPods := map[string][]*v1.Pod{}
+	var resultLock sync.Mutex
+
+	// We can use the same metadata producer for all nodes.
+	meta := metadataProducer(pod, nodeNameToInfo)
+	checkNode := func(i int) {
+		nodeName := nodes[i].Name
+		pods, fits := selectVictimsOnNode(pod, meta, nodeNameToInfo[nodeName], predicates)
+		if fits {
+			resultLock.Lock()
+			nodeNameToPods[nodeName] = pods
+			resultLock.Unlock()
+		}
+	}
+	workqueue.Parallelize(16, len(nodes), checkNode)
+	return nodeNameToPods
+}
+
+// selectVictimsOnNode finds minimum set of pods on the given node that should
+// be preempted in order to make enough room for "pod" to be scheduled.
+// The algorithm first checks if the pod can be scheduled on the node when all the
+// lower priority pods are gone. If so, it sorts all the lower priority pods by
+// their priority and starting from the highest priority one, tries to keep as
+// many of them as possible while checking that the "pod" can still fit on the node.
```

Comment:
BTW I know you have been thinking about how to handle the pod affinity case, but I'm really wondering whether it is necessary for the first version. IIUC, the algorithm you are using will still "work" in the presence of pod affinity, in the same sense that it "works" for anti-affinity when the TopologyKey is larger than a node -- namely, it may say a pod is infeasible everywhere when actually it could be scheduled somewhere (with preemption). But we can explain those circumstances in the release notes and fix it in the next version, I think. (Of course you should make sure this is actually true, and that you can precisely describe the conditions where it will act in this sub-optimal way. Defining those circumstances is a pre-requisite for fixing the issue anyway.)